src/interfaces/swig/shogun.i

```diff
@@ -34,16 +34,357 @@
+					handle_pyarray(v);
+				else if (*m_py_obj && PyList_Check(*m_py_obj))
+					handle_pylist(v);
+				else
+				  *m_py_obj = sg_to_python(v);
+			}
+
+			template <typename T>
+			PyObject* sg_to_python(const T* v)
+			{
+				if constexpr(std::is_same_v<T, bool>)
+				{
+					PyObject* result = *v ? Py_True : Py_False;
+					Py_INCREF(result);
+					return result;
+				}
+				if constexpr(std::is_same_v<T, int8_t>)
+					return PyLong_FromLong(static_cast<long>(*v));
+				if constexpr(std::is_same_v<T, int16_t>)
+					return PyLong_FromLong(static_cast<long>(*v));
+				if constexpr(std::is_same_v<T, int32_t>)
+					return PyLong_FromLong(*v);
+				if constexpr(std::is_same_v<T, int64_t>)
+					return PyLong_FromLongLong(*v);
+				if constexpr(std::is_same_v<T, float32_t>)
+					return PyFloat_FromDouble(static_cast<double>(*v));
+				if constexpr(std::is_same_v<T, float64_t>)
+					return PyFloat_FromDouble(static_cast<double>(*v));
+				if constexpr(std::is_same_v<T, floatmax_t>)
+					return PyFloat_FromDouble(static_cast<double>(*v));
+				if constexpr(std::is_same_v<T, char>)
+					return SWIG_FromCharPtr((const char *)v);
+				if constexpr(std::is_same_v<T, uint8_t>)
+					return PyLong_FromUnsignedLong(static_cast<unsigned long>(*v));
+				if constexpr(std::is_same_v<T, uint16_t>)
+					return  PyLong_FromUnsignedLong(static_cast<unsigned long>(*v));
+				if constexpr(std::is_same_v<T, uint32_t>)
+					return PyLong_FromSize_t(static_cast<size_t>(*v));
+				if constexpr(std::is_same_v<T, uint64_t>)
+					return PyLong_FromUnsignedLong(static_cast<size_t>(*v));
+				if constexpr(std::is_same_v<T, complex128_t>)
+					return PyComplex_FromDoubles(v->real(), v->imag());
+				if constexpr(std::is_same_v<T, CSGObject*>)
+					return SWIG_InternalNewPointerObj(SWIG_as_voidptr(*v), SWIGTYPE_p_shogun__CSGObject, 0);
+				error("Cannot handle casting from shogun type {} to python type!", demangled_type<T>().c_str());
+			}
+
+			template <typename T>
+			void handle_pyarray(const T* v)
+			{
+				if constexpr(std::is_same_v<T, char>)
+				{
+					if (!(*m_py_obj))
+						*m_py_obj = PyUnicode_FromString(v);
+				}
+				else
+				{
+					if (dims && !(*m_py_obj))
+					{
+						PyArray_Descr* descr=PyArray_DescrFromType(sg_to_npy_type<T>::type);
+						*m_py_obj = PyArray_NewFromDescr(&PyArray_Type,
+							descr, n_dims, dims, NULL, NULL,  NPY_ARRAY_FARRAY | NPY_ARRAY_WRITEABLE, NULL);
+						PyArray_ENABLEFLAGS((PyArrayObject*) m_py_obj, NPY_ARRAY_OWNDATA);
+					}
+					else if (!dims && !(*m_py_obj))
+					{
+						return;
+					}
+					((T*)PyArray_DATA((PyArrayObject*)*m_py_obj))[current_i] = *v;
+				}
+				current_i++;
+			}
+
+			template <typename T>
+			void handle_pylist(const T* v)
+			{
+				bool new_obj = m_nested_py_obj ? false : true;
+				auto nested_visitor = PythonVisitor(m_nested_py_obj, dims, nested_current_i, n_dims);
+				nested_visitor.on(const_cast<T*>(v));
```

Comment:
@karlnapf I think the visitor should be always const pointer, i.e. `on(const T*)`. I dont think there is a situation where you would change the pointer in the visitor?

Reply:
@gf712 const T* doesn't work because actually in some circumstances with the visitor you set the values (see deserialization)... in this case it works of course but the base interface AnyVisitor::on should be T*

Reply:
ah makes sense, i was wondering why they weren't const